(feature) Write Vacancy publish events to a Google spreadsheet with Sidekiq

lozette opened the pull request:
## Trello card URL:

https://trello.com/c/71M4SnCU/522-vacancy-publish-events-to-be-added-to-google-sheets-using-sidekiq-instead-of-papertrail-and-zapier

## Changes in this PR:

- Vacancy#to_row added to create a CSV row representation of the published vacancy
- Sidekiq job created and added to new queue, to be called when vacancy is published
- Spreadsheet::Writer repurposed to write a single row at a time to the Vacancy spreadsheet
- `audit.rake` deleted as it is no longer used, and Spreadsheet::Writer no longer supports it

(Out of scope of this PR)

- Spreadsheet created in DfE Google suite, id `15KTqvwNZWPaDPSzGaoClYh0d2jUWz4C5mpwn7jLmt2E`
https://docs.google.com/spreadsheets/d/15KTqvwNZWPaDPSzGaoClYh0d2jUWz4C5mpwn7jLmt2E/edit?ts=5bd9c1b9#gid=0
- staging Google account service writer added as a writer to this spreadsheet

## Next steps:

- [ ] `PUBLISHED_VACANCY_SPREADSHEET_ID` value to be added to local environment set ups
- [ ] `PUBLISHED_VACANCY_SPREADSHEET_ID` value to be added to Terraform env vars
- [ ] Spreadsheet created in DfE Google suite for production Vacancies,
- [ ] id for production spreadsheet added to production Terraform vars
- [ ] production Google account service writer (production-writer@teacher-vacancy-service.iam.gserviceaccount.com) added as a writer to this spreadsheet

lozette commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Vacancy publish events to be added to Google Sheets using Sidekiq instead of Papertrail and Zapier](https://trello.com/c/71M4SnCU/522-vacancy-publish-events-to-be-added-to-google-sheets-using-sidekiq-instead-of-papertrail-and-zapier)

despo commented on app/jobs/update_vacancy_spreadsheet_job.rb:
```diff
@@ -0,0 +1,22 @@
+class UpdateVacancySpreadsheetJob < ApplicationJob
+  queue_as :update_vacancy_spreadsheet
+
+  PUBLISHED_VACANCY_SPREADSHEET_ID = ENV['PUBLISHED_VACANCY_SPREADSHEET_ID']
```
Can this be moved in an initialiser, so all relevant configuration is in one place.

[The google_keys initializer]( https://github.com/dxw/teacher-vacancy-service/blob/develop/config/initializers/google_keys.rb) looks like a good candidate as it's only used by the `SpreadsheetWriter`. The initializer should probably be renamed to reflect that as well.

despo commented on app/models/vacancy.rb:
```diff
@@ -165,6 +165,23 @@ def refresh_slug
     send(:set_slug)
   end
 
+  def to_row
```
Can we move this and the spec to the `VacancyPresenter`

despo commented on spec/models/vacancy_spec.rb:
```diff
@@ -433,4 +433,18 @@
   end
+
+  describe '#to_row' do
+    let(:vacancy) { build(:vacancy) }
```
You should be able to use `create(:vacancy)` instead of `build` and `save!` in the before block as the default factory has all required values set.

despo commented on terraform/container-definitions/web_container_definition.json:
```diff
@@ -126,6 +126,10 @@
+      {
+        "name": "PUBLISHED_VACANCY_SPREADSHEET_ID",
+        "value": "${published_vacancy_spreadsheet_id}"
+      },
```
This needs to be configured in the `worker_container_definition` as that is where the jobs are executed.

despo commented on terraform/modules/ecs/ecs.tf:
```diff
@@ -157,6 +157,7 @@ data "template_file" "web_container_definition" {
     pp_user_satisfaction_token       = "${var.pp_user_satisfaction_token}"
     google_drive_json_key            = "${var.google_drive_json_key}"
     auth_spreadsheet_id              = "${var.auth_spreadsheet_id}"
+    published_vacancy_spreadsheet_id = "${var.published_vacancy_spreadsheet_id}"
```
This needs to be moved to the worker_container_definition template.

lozette commented:
Vacancy written to spreadsheet after testing on edge:

https://docs.google.com/spreadsheets/d/15KTqvwNZWPaDPSzGaoClYh0d2jUWz4C5mpwn7jLmt2E/edit?ts=5bd9c1b9#gid=0

I have amended the `google_drive_json_key` values in `staging.tfvars` and `edge.tfvars` to be the new "map" format.

lozette commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Vacancy publish events to be added to Google Sheets using Sidekiq instead of Papertrail and Zapier](https://trello.com/c/71M4SnCU/522-vacancy-publish-events-to-be-added-to-google-sheets-using-sidekiq-instead-of-papertrail-and-zapier)

despo commented on terraform.tf:
```diff
@@ -121,6 +121,7 @@ module "ecs" {
   authorisation_service_url        = "${var.authorisation_service_url}"
   google_drive_json_key            = "${var.google_drive_json_key}"
   auth_spreadsheet_id              = "${var.auth_spreadsheet_id}"
```
This variable is no longer used, so can we remove it please.

despo commented on terraform/container-definitions/web_container_definition.json:
```diff
@@ -126,6 +126,10 @@
         "name": "AUTH_SPREADSHEET_ID",
         "value": "${auth_spreadsheet_id}"
       },
```
container definition entries should be removed as well.

lozette commented:
Removed `auth_spreadsheet_id` from code and all `.tfvars` in 1password

Also added the production published vacancies spreadsheet id to `production.tfvars` in 1password

tahb commented:
@lozette Looks like this is ready to go short of a rebase 👍

lozette commented:
Spreadsheet URLs:

Production: https://docs.google.com/spreadsheets/d/1LT4GSXnaBJRYBlaJToYdt2JyQ78fKk5qw7jCXxR_kY8/edit#gid=0

Staging/edge: https://docs.google.com/spreadsheets/d/1SDJxhphZ_337aZTR515AOLuldOwEjnG6T1_PVYLuQE0/edit#gid=0
